# Retail Intelligence Project
## Data Understanding

Goal:
Understand the structure and quality of the retail sales dataset before performing any transformations.

Steps in this notebook:
1. Load dataset
2. Explore dataset structure
3. Identify missing values
4. Check duplicates
5. Generate summary statistics
6. Document observations

In [2]:
## Import Libraries

import pandas as pd
import numpy as np


In [4]:
file_path  = r"D:\Analyst\DigitalProduct\Automated_Retail_Intelligence\data\raw\online_retail.csv"
df = pd.read_csv(file_path)
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [6]:
df.shape

(1067371, 8)

In [8]:
df.columns

Index(['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'Price', 'Customer ID', 'Country'],
      dtype='object')

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1067371 non-null  object 
 1   StockCode    1067371 non-null  object 
 2   Description  1062989 non-null  object 
 3   Quantity     1067371 non-null  int64  
 4   InvoiceDate  1067371 non-null  object 
 5   Price        1067371 non-null  float64
 6   Customer ID  824364 non-null   float64
 7   Country      1067371 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 65.1+ MB


In [11]:
df.isnull().sum()

Invoice             0
StockCode           0
Description      4382
Quantity            0
InvoiceDate         0
Price               0
Customer ID    243007
Country             0
dtype: int64

In [12]:
df.duplicated().sum()

np.int64(34335)

In [17]:
df["Country"].nunique()

43

In [18]:
df["Customer ID"].nunique()

5942

In [19]:
df[df["Quantity"] < 0].head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
178,C489449,22087,PAPER BUNTING WHITE LACE,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
179,C489449,85206A,CREAM FELT EASTER EGG BASKET,-6,2009-12-01 10:33:00,1.65,16321.0,Australia
180,C489449,21895,POTTING SHED SOW 'N' GROW SET,-4,2009-12-01 10:33:00,4.25,16321.0,Australia
181,C489449,21896,POTTING SHED TWINE,-6,2009-12-01 10:33:00,2.10,16321.0,Australia
182,C489449,22083,PAPER CHAIN KIT RETRO SPOT,-12,2009-12-01 10:33:00,2.95,16321.0,Australia


In [20]:
df[df["Price"] <= 0].head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
263,489464,21733,85123a mixed,-96,2009-12-01 10:52:00,0.0,NaN,United Kingdom
283,489463,71477,short,-240,2009-12-01 10:52:00,0.0,NaN,United Kingdom
284,489467,85123A,21733 mixed,-192,2009-12-01 10:53:00,0.0,NaN,United Kingdom
470,489521,21646,NaN,-50,2009-12-01 11:44:00,0.0,NaN,United Kingdom
3114,489655,20683,NaN,-44,2009-12-01 17:26:00,0.0,NaN,United Kingdom


In [21]:
df[df["Invoice"].str.contains("C", na=False)].head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
178,C489449,22087,PAPER BUNTING WHITE LACE,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
179,C489449,85206A,CREAM FELT EASTER EGG BASKET,-6,2009-12-01 10:33:00,1.65,16321.0,Australia
180,C489449,21895,POTTING SHED SOW 'N' GROW SET,-4,2009-12-01 10:33:00,4.25,16321.0,Australia
181,C489449,21896,POTTING SHED TWINE,-6,2009-12-01 10:33:00,2.10,16321.0,Australia
182,C489449,22083,PAPER CHAIN KIT RETRO SPOT,-12,2009-12-01 10:33:00,2.95,16321.0,Australia


In [32]:
df["InvoiceDate"].min(), df["InvoiceDate"].max()

('2009-12-01 07:45:00', '2011-12-09 12:50:00')

# Data Understanding – Key Observations

## 1. Dataset Overview

The dataset contains **1,067,371 rows and 8 columns**, representing transactional retail sales data.

Each row represents a **product line within an invoice**, meaning a single invoice can contain multiple products purchased by a customer.

The columns present in the dataset are:

* Invoice – Invoice number for each transaction
* StockCode – Unique product identifier
* Description – Product name or description
* Quantity – Number of items purchased
* InvoiceDate – Date and time of the transaction
* Price – Unit price of the product
* Customer ID – Unique customer identifier
* Country – Country where the customer is located

---

## 2. Data Types

The dataset contains three types of data fields:

* **Categorical/Text:** Invoice, StockCode, Description, Country
* **Numerical:** Quantity, Price
* **Customer Identifier:** Customer ID

One important issue identified is that **InvoiceDate is stored as an object instead of a datetime datatype**, which will need to be converted during the data cleaning stage for proper time-based analysis.

---

## 3. Missing Values

Two columns contain missing values:

| Column      | Missing Records |
| ----------- | --------------- |
| Description | 4,382           |
| Customer ID | 243,007         |

Observations:

* Missing values in **Description** are relatively small compared to the dataset size.
* A significant number of records have **missing Customer IDs (~22%)**, which suggests that some purchases were made by **guest customers or unidentified buyers**.

For customer-level analysis (such as repeat purchase behaviour or RFM analysis), rows with missing Customer IDs may need to be excluded.

---

## 4. Duplicate Records

The dataset contains **34,335 duplicate rows**.

Possible reasons include:

* repeated invoice entries
* system duplication
* multiple identical product entries recorded more than once

These duplicates will need to be **investigated and removed during data cleaning** to ensure accurate analysis.

---

## 5. Customer Distribution

The dataset contains **5,942 unique customers**.

This indicates that customer-level analysis such as:

* repeat purchase behaviour
* customer lifetime value
* customer segmentation

can be performed.

---

## 6. Geographic Coverage

The dataset contains customers from **43 different countries**, indicating that the retailer operates internationally.

However, the majority of transactions are likely concentrated in a few primary markets (commonly the United Kingdom in this dataset).

---

## 7. Product Returns and Cancellations

Several records contain **negative quantities**.

Example observation from the dataset:

* Negative quantities appear in invoices that begin with the letter **"C"**.

This indicates that these transactions represent **order cancellations or product returns**.

Thus:

* **Invoice numbers starting with "C" represent cancelled orders**
* Negative quantities correspond to returned items.

These transactions must be handled carefully during analysis to avoid incorrect revenue calculations.

---

## 8. Zero Price Transactions

Some transactions contain **Price = 0**.

These may represent:

* promotional items
* free samples
* system adjustments
* data entry anomalies

These records will need to be examined during the cleaning stage before revenue calculations are performed.

---

## 9. Date Range

Date ranging from start date of Invoice to end date of Invoice in the dataset

* Start date : 2009-12-01 
* End date : 2011-12-09

## 10. Dataset Characteristics

Based on the initial exploration, the dataset represents a **large-scale retail transactional system** with the following characteristics:

* Line-level transaction records
* Multi-product invoices
* International customer base
* Product returns and cancellations
* Potential duplicate records

This makes the dataset suitable for building a **Retail Intelligence and Sales Analytics dashboard**.

---

## 11. Data Issues Identified for Cleaning

The following data quality issues were identified and will be addressed in the next stage:

1. Convert **InvoiceDate to datetime format**
2. Remove **duplicate records**
3. Handle **missing Customer IDs**
4. Investigate **missing product descriptions**
5. Separate **cancelled transactions (Invoice starting with 'C')**
6. Handle **negative quantities (returns)**
7. Investigate **zero price transactions**

These steps will be performed in the **Data Cleaning notebook**.


In [33]:
df["Quantity"].min(), df["Quantity"].max()

(np.int64(-80995), np.int64(80995))

In [34]:
df["Price"].min(), df["Price"].max()

(np.float64(-53594.36), np.float64(38970.0))